### ISSUE TITLE
Feature Request: Minimum Consecutive Hours for Split

### ISSUE THREAD

**Issue by sembeek** [TheFes/cheapest-energy-hours#160]
Hi,

First of all, thank you for this fantastic project! It's been incredibly helpful for a variety of tasks in my setup.

One feature I'd love to see (or perhaps it already exists, and I haven't figured it out) is related to calculating the 8 cheapest hours for my heat pump. Specifically, I'd like to be able to "split" these hours, so that the system can schedule, for example, 4 cheaper hours at night and 4 in the afternoon.

However, with the current implementation, I’ve noticed that the system sometimes selects short periods, such as a single hour, before stopping, which isn’t ideal for a heat pump since it requires longer operational periods.

Would it be possible to add an option that enforces a minimum number of consecutive hours when splitting? This would ensure the heat pump runs efficiently without frequent starts and stops.

Looking forward to your thoughts!

Best regards,
Sem

**Comment by rrooggiieerr** [comment]
I'm quite sure you can use the `hours` parameter to set the number consecutive hours to be used, thus:

```
hours=4
```

**Comment by sembeek** [comment]
That will just specifiy for how many hours it will search right?
In my experience it is possible that hours=4 can choose 2 hours in the night and 2 in the afternoon, in combination with split.

**Comment by rrooggiieerr** [comment]
If you look at this example from the documentation I would think it's about the duration
```
{% from 'cheapest_energy_hours.jinja' import cheapest_energy_hours %}
{% set output = cheapest_energy_hours(sensor='sensor.cheap_energy', hours=2.25, start='12:00', end=today_at('19:00') + timedelta(days=1), look_ahead=true) %}
```
But that's very easy to test in Developer Tools > Template

**Comment by rrooggiieerr** [comment]
Actually, I'm using the `hours` parameter in one of my automations and it works exactly like your usecase

**Comment by sembeek** [comment]
I understand what you are saying, but you are not using the split parameter.
I would like to use the split parameter so the hours can be split in day and night if that is cheaper.

**Comment by rrooggiieerr** [comment]
My suggestion would be to have your feature request to be to allow integer values for the `split` parameter where `true` and `false` (default) would work as currently implemented and an integer value would return a maximum number of splits. `split=2` would then give you the 2 periods within the day that are cheapest

**Comment by KrzysztofHajdamowicz** [comment]
I'm not sure if it's something worth pursuit.

Just look for 4 hours between 02:00 and 12:00 and 4 hours between 14:00 and midnight or something like this.

You can compare current hour to decide which start and end values are given to macro so you can still use single binary sensor entity.

**Comment by sembeek** [comment]
Yeah but that would not work if the 8 cheapest hours are between 10:00 and 18:00 for example.

**Comment by KrzysztofHajdamowicz** [comment]
But is single 8h run enough for heating home during 24h window?

I bet single 4h run provides heat for next 12h and running pump longer won't help as heat reservoir is already full.

Also, as heat pump prefer longer runs where it can just keep heat reservoir topped-off, I'd suggest to search 2-4 most expensive hours and disable heat pump during that time.
It will give long cycle times and better thermal comfort.

**Comment by sembeek** [comment]
Single 8h run is enough for my home, i have a really well isolated small appartment, so it does not need much.
I still have a home assistant thermostat attached to it, so it won't turn on if it already met the temperature.
With a lot of tinkering and monitoring i found out that around 8 hours a day is the most efficient for me, it does not matter if it is splitted in 2 times 4 hours or 8 hours straight.

**Comment by TheFes** [comment]
I understand the request, you basically want a minimum period length in case `split=true` is used
I'm kinda busy at the moment, but I will try to find some time to implement this.

**Comment by tvdbroeck** [comment]
The minimum period for split=True has become more relevant with the 15min prices. 
Right now it will output short periods of 15min too. 
It would be helpful to be able to select eg the 4 cheapest hours with a minimum duration of eg 2 or 4 15min intervals. This to avoid waking the heat pump for only 15 minutes.

**Comment by TheFes** [comment]
I'm thinking about how to implement this, but it is rather complex.
Let's say you want 2 hours minimum per block, and you need 5 hours in total.
I then need to find a way to use 2 hour sections, it should not be limited to that, because it could be that a 2.75 and 2.25 split is optimum, while with splitting in 2 hour sections it will lead to a 2, 2, 1 split (which could be combined in something like 2 and 3, or even 5)
With a lower minimum number of hours and/or a larger total number of hours, the number of possible consecutive datapoints becomes even bigger.

I need some time to figure out the best way to do this, I'm open for suggestions :)

**Comment by TheFes** [comment]
Thanks, but this is Python code, not Jinja.
But I'll see if I understand the logic in the code and if I can apply it to the macro

**Comment by TheFes** [comment]
v7.2.0 is now released, closing this issue.
Please create a new issue in case there are problems.
